_sources/Unit1-Using-Objects-and-Methods/topic-1-8-comments.rst

```diff
@@ -73,53 +73,87 @@ Notice that there are some special tags that you can use in Java documentation.
+Preconditions and Postconditions
+----------------------------------
+
+Many methods in API libraries have **preconditions** and the **postconditions** described in their comments. A **precondition** is a condition that must be true for your method code to work, for example the assumption that the parameters, the data that you give the method to do its job, have values within limits and are not null. The methods could check for these preconditions, but they do not have to. The precondition is what the method expects in order to do its job properly.
 
-As you write methods in a class, it is a good idea to keep in mind the **preconditions** and the **postconditions** for the method and write them in the comments. A precondition is a condition that must be true for your method code to work, for example the assumption that the parameters have values and are not null. The methods could check for these preconditions, but they do not have to. The precondition is what the method expects in order to do its job properly.
+A **postcondition** is a condition that is true after running the method. It is what the method promises to do. Postconditions describe the outcome of running the method, for example what is being returned from the method or the changes to the state. These assumptions are very useful to other programmers who want to use that method and get the correct results.
 
-A postcondition is a condition that is true after running the method. It is what the method promises to do. Postconditions describe the outcome of running the method, for example what is being returned or the changes to the instance variables. These assumptions are very useful to other programmers who want to use your class and get the correct results.
+One precondition that we talked about is that divisors cannot be zero in expressions because that will cause the code to have a runtime error or exception. Many math functions have preconditions about their operands. For example, computing the square root of a negative number is undefined, so the ``Math.sqrt(num)`` Java method which we will learn later will throw an exception (a runtime error) if num is negative. These are described as special cases in the API documentation, for example in https://docs.oracle.com/javase%2F8%2Fdocs%2Fapi%2F%2F/java/lang/Math.html#sqrt-double-. Try this out below.
 
+|CodingEx| **Coding Exercise**
 
-Here is an example of preconditions, postconditions, and @param in the Turtle code that we have used in the past for our drawing turtles.
+.. activecode:: math-preconditions
+   :language: java
+   :autograde: unittest
+
+   The following code uses the square root method in Java which has a precondition that the number that you give it is not negative.  When you click on run, the compiler will not catch the error, but there will be a run-time error or exception.  Can you fix the value of num so that it does not throw an exception?  What is the precondition for the Math.sqrt method?
```

Comment:
```suggestion
   The following code uses the square root method in Java which has a precondition that the number that you give it is not negative.  When you click on run, the compiler will not catch the error, but it will return the error value ``NaN``.  Can you fix the value of num so that it does not return ``NaN``?  What is the precondition for the Math.sqrt method?
```